bcachefs/bcachefs.c

```diff
@@ -4,42 +4,108 @@
 
 #include "bcachefs.h"
 
+// Our data structure structs are really just header of contiguous lists.  Most
+// of the time, the header always start with the size of full list in bytes
+//
+// The elements are starting after the header struct, they can have different
+// size, their size is also stored in the first few bytes of the element u64s
+// can have different sizes (uint8_t to uint64_t), to know the number of bytes
+// used to store the size we need to use the `struct u64s_spec`
+//
+//  * struct header {
+//  |  u64s                // Size of the entire data structure
+//  |  metadata
+//  |       .
+//  |       .
+//  |       .
+//  |  metadata   };       //
+//  | *  struct Value {
+//  | |   u64s              // Size of this value in `BCH_U64S_SIZE`
+//  | +> };
+//  | *  struct Value {
+//  | |   u64s              // Size of this value in `BCH_U64S_SIZE`
+//  | +> };
+//  +>                      // End of data structure
+//
+//  end         = &header + header->u64s;
+//  first_value = &header + sizeof(header);
+
+
+// Reads the u64s field contained inside a struct (assume it is the first
+// field) note that fields like `uint64_t _data[0];` do not contribute to the
+// struct size and create a pointer to the begining of the struct.
+//
+// TODO: duplicate with benz_uintXX_as_uint64
+uint64_t read_u64s(const void *c, struct u64s_spec u64s_spec) {
+    uint64_t u64s = 0;
+    switch (u64s_spec.size)
+    {
+    case sizeof(uint8_t):
+        u64s = *((const uint8_t*)c);
+        break;
+    case sizeof(uint16_t):
+        u64s = *((const uint16_t*)c);
+        break;
+    case sizeof(uint32_t):
+        u64s = *((const uint32_t*)c);
+        break;
+    case sizeof(uint64_t):
+        u64s = *((const uint64_t*)c);
+        break;
+    default:
+        u64s = 0;
+    }
+    return u64s;
+}
+
+// Gets next element, reads the size of the current element and jump to the
+// next one
+//
+// if current element is null, set it to the first element
+// if current element is higher than the end returns null
+//
+// p            : start of our parent data structure
+// size_of_p    : size of the header of our parent data structure
+// p_end        : end of the elements
+// c            : current child element
+// u64s_spec    : number of bytes used to store the element size
+//
+// NOTE: benz_bch_next_sibling is currently used in 3 locations and everytime
+// p_end is `p + p->u64s * BCH_U64S_SIZE`
+//
+// we could simplify this function as
+//
+//  benz_bch_next_sibling(const void* start, const void* end, const void* elem, u64s_spec u64s_spec);
+//
+// with:
+//  start = p + sizeof(p)
+//  end = p + p->u64s * BCH_U64S_SIZE
+//
```

Comment:
`sizeof_p` will always be customizable in the caller so we are not losing much in terms of visibility.
Instead of doing the addition in the body of the loop we will be doing outside which in turn will make the initialization of p_start & p_end close together.

I think ultimately it would make it easier to read.

Reply:
It's only a bit annoying to have to cast to (const uint8_t*) before advancing the pointer but if it's only needed in prototyping situations that's not a big deal. I would think we should aim to have all `*_next_*()` to follow a similar signature as much as possible (parent `p` and child `c` is found in all of them):
```
const void *benz_bch_next_sibling(const void *p, uint32_t sizeof_p, const void *p_end, const void *c, struct u64s_spec u64s_spec);

const struct bch_sb_field *benz_bch_next_sb_field(const struct bch_sb *p, const struct bch_sb_field *c, enum bch_sb_field_type type);
const struct jset_entry *benz_bch_next_jset_entry(const struct bch_sb_field *p,
                                                  uint32_t sizeof_p,
                                                  const struct jset_entry *c,
                                                  enum bch_jset_entry_type type);
const struct bch_val *benz_bch_next_bch_val(const struct bkey *p, const struct bch_val *c, uint32_t sizeof_c);
const struct bset *benz_bch_next_bset(const struct btree_node *p, const struct bset *c, const struct bch_sb *sb);
const struct bkey *benz_bch_next_bkey(const struct bset *p, const struct bkey *c, enum bch_bkey_type type);
```

At the same time, the structures might be too different one from another so that might be a bit of a lost cause. I also see a bit of auto doc by having the `sizeof_p` explicit. If we use a `(const void* start, const void* end, ...)` signature the caller needs to know that he has to add some value to the parent `p` before sending it to the the function. But in the end, I would not oppose to a PR that includes the change. Maybe this function could also be split into 2:
```
benz_bch_first_child()
benz_bch_next_sibling()
```